server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java

```diff
@@ -33,20 +32,25 @@
 
     private final List<FieldAndFormat> fields;
 
-    public static FetchDocValuesContext create(Function<String, Set<String>> simpleMatchToFullName,
-                                               int maxAllowedDocvalueFields,
-                                               List<FieldAndFormat> fieldPatterns) {
+    public static FetchDocValuesContext create(MapperService mapperService, List<FieldAndFormat> fieldPatterns) {
```

Comment:
I think this signature change undoes some of @javanna's work in https://github.com/elastic/elasticsearch/pull/63239. The easiest fix would be to keep the function parameter, and add a new one `QueryShardContext::isFieldMapped`.

As a side note, if we end up passing a lot of functions to avoid exposing `MapperService`, I wonder if we should pull out a lightweight abstraction similar to `FieldTypeLookup` and pass it around instead?

Reply:
thanks for catching this Julie. I am still addressing the original problem: we carry MapperService around too often when it's not needed. I do see the increasing need for functions which is definitely a signal that we dont have the right abstractions. I am using functions in some places but I see it more as a temporary work-around. I do think once we remove getMapperService from QueryShardContext we should reconsider the functions needed. A field type lookup abstraction may also be needed as I would like to split DocumentMapper into some sub-components that can be shared independently. To be seen very soon :)

Reply:
Instead of using MapperService, I pass in QueryShardContext now which has methods to resolve field name patterns, allows looking up the setting and look up whether a field is mapped. I can go back to passing in three parameters (functions and the setting itself) if you think this is not ideal for any reason.